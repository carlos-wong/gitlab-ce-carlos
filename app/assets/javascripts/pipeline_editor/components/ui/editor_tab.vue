<script>
import { GlAlert, GlBadge, GlTab } from '@gitlab/ui';
import { __, s__ } from '~/locale';
/**
 * Wrapper of <gl-tab> to optionally lazily render this tab's content
 * when its shown **without dismounting after its hidden**.
 *
 * Usage:
 *
 * API is the same as <gl-tab>, for example:
 *
 * <gl-tabs>
 *   <editor-tab title="Tab 1" lazy>
 *     lazily mounted content (gets mounted if this is first tab)
 *   </editor-tab>
 *   <editor-tab title="Tab 2" lazy>
 *     lazily mounted content
 *   </editor-tab>
 *   <editor-tab title="Tab 3">
 *      eagerly mounted content
 *   </editor-tab>
 * </gl-tabs>
 *
 * Once the tab is selected it is permanently set as "not-lazy"
 * so it's contents are not dismounted.
 *
 * lazy is "false" by default, as in <gl-tab>.
 *
 * It is also possible to pass the `isEmpty` and or `isInvalid` to let
 * the tab component handle that state on its own. For example:
 *
 *  * <gl-tabs>
 *   <editor-tab-with-status title="Tab 1" :is-empty="isEmpty" :is-invalid="isInvalid">
 *    ...
 *   </editor-tab-with-status>
 *   Will be the same as normal, except it will only render the slot component
 *   if the status is not empty and not invalid. In any of these 2 cases, it will render
 *   a generic component and avoid mounting whatever it received in the slot.
 * </gl-tabs>
 */

export default {
  i18n: {
    invalid: __('Your CI/CD configuration syntax is invalid. View Lint tab for more details.'),
    unavailable: __(
      "We're experiencing difficulties and this tab content is currently unavailable.",
    ),
  },
  components: {
    GlAlert,
    GlBadge,
    GlTab,
    // Use a small renderless component to know when the tab content mounts because:
    // - gl-tab always gets mounted, even if lazy is `true`. See:
    // https://github.com/bootstrap-vue/bootstrap-vue/blob/dev/src/components/tabs/tab.js#L180
    // - we cannot listen to events on <slot />
    MountSpy: {
      render: () => null,
    },
  },
  inheritAttrs: false,
  props: {
    badgeTitle: {
      type: String,
      required: false,
      default: '',
    },
    badgeVariant: {
      type: String,
      required: false,
      default: 'info',
    },
    emptyMessage: {
      type: String,
      required: false,
      default: s__(
        'PipelineEditor|This tab will be usable when the CI/CD configuration file is populated with valid syntax.',
      ),
    },
    isEmpty: {
      type: Boolean,
      required: false,
      default: false,
    },
    isInvalid: {
      type: Boolean,
      required: false,
      default: false,
    },
    isUnavailable: {
      type: Boolean,
      required: false,
      default: false,
    },
    keepComponentMounted: {
      type: Boolean,
      required: false,
      default: true,
    },
    lazy: {
      type: Boolean,
      required: false,
      default: false,
    },
    title: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      isLazy: this.lazy,
    };
  },
  computed: {
    hasBadgeTitle() {
      return this.badgeTitle.length > 0;
    },
    slots() {
      // eslint-disable-next-line @gitlab/vue-prefer-dollar-scopedslots
      return Object.keys(this.$slots);
    },
  },
  methods: {
    onContentMounted() {
      // When a child is first mounted make the entire tab
      // permanently mounted by setting 'lazy' to false unless
      // explicitly opted out.
      if (this.keepComponentMounted) {
        this.isLazy = false;
      }
    },
  },
};
</script>
<template>
  <gl-tab :lazy="isLazy" v-bind="$attrs" v-on="$listeners">
    <template #title>
      <span>{{ title }}</span>
      <gl-badge v-if="hasBadgeTitle" class="gl-ml-2" size="sm" :variant="badgeVariant">{{
        badgeTitle
      }}</gl-badge>
    </template>
    <gl-alert v-if="isEmpty" variant="tip">{{ emptyMessage }}</gl-alert>
    <gl-alert v-else-if="isUnavailable" variant="danger" :dismissible="false">
      {{ $options.i18n.unavailable }}</gl-alert
    >
    <gl-alert v-else-if="isInvalid" variant="danger">{{ $options.i18n.invalid }}</gl-alert>
    <template v-else>
      <slot v-for="slot in slots" :name="slot"></slot>
      <mount-spy @hook:mounted="onContentMounted" />
    </template>
  </gl-tab>
</template>
