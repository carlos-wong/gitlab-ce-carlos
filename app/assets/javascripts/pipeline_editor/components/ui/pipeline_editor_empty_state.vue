<script>
import { GlButton, GlSprintf } from '@gitlab/ui';
import { __ } from '~/locale';
import PipelineEditorFileNav from '~/pipeline_editor/components/file_nav/pipeline_editor_file_nav.vue';

export default {
  components: {
    GlButton,
    GlSprintf,
    PipelineEditorFileNav,
  },
  i18n: {
    title: __('Optimize your workflow with CI/CD Pipelines'),
    body: __(
      'Create a new %{codeStart}.gitlab-ci.yml%{codeEnd} file at the root of the repository to get started.',
    ),
    btnText: __('Configure pipeline'),
  },
  inject: {
    emptyStateIllustrationPath: {
      default: '',
    },
  },
  methods: {
    createEmptyConfigFile() {
      this.$emit('createEmptyConfigFile');
    },
  },
};
</script>
<template>
  <div>
    <pipeline-editor-file-nav v-on="$listeners" />
    <div class="gl-display-flex gl-flex-direction-column gl-align-items-center gl-mt-11">
      <img :src="emptyStateIllustrationPath" />
      <h1 class="gl-font-size-h1">{{ $options.i18n.title }}</h1>
      <p class="gl-mt-3">
        <gl-sprintf :message="$options.i18n.body">
          <template #code="{ content }">
            <code>{{ content }}</code>
          </template>
        </gl-sprintf>
      </p>
      <gl-button
        variant="confirm"
        class="gl-mt-3"
        data-qa-selector="create_new_ci_button"
        @click="createEmptyConfigFile"
      >
        {{ $options.i18n.btnText }}
      </gl-button>
    </div>
  </div>
</template>
