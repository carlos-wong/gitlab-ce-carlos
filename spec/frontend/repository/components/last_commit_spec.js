import { shallowMount } from '@vue/test-utils';
import { GlLoadingIcon } from '@gitlab/ui';
import LastCommit from '~/repository/components/last_commit.vue';
import UserAvatarLink from '~/vue_shared/components/user_avatar/user_avatar_link.vue';

let vm;

function createCommitData(data = {}) {
  const defaultData = {
    sha: '123456789',
    title: 'Commit title',
    message: 'Commit message',
    webUrl: 'https://test.com/commit/123',
    authoredDate: '2019-01-01',
    author: {
      name: 'Test',
      avatarUrl: 'https://test.com',
      webUrl: 'https://test.com/test',
    },
    pipeline: {
      detailedStatus: {
        detailsPath: 'https://test.com/pipeline',
        icon: 'failed',
        tooltip: 'failed',
        text: 'failed',
        group: {},
      },
    },
  };
  return Object.assign(defaultData, data);
}

function factory(commit = createCommitData(), loading = false) {
  vm = shallowMount(LastCommit, {
    mocks: {
      $apollo: {
        queries: {
          commit: {
            loading: true,
          },
        },
      },
    },
  });
  vm.setData({ commit });
  vm.vm.$apollo.queries.commit.loading = loading;
}

const emptyMessageClass = 'font-italic';

describe('Repository last commit component', () => {
  afterEach(() => {
    vm.destroy();
  });

  it.each`
    loading  | label
    ${true}  | ${'shows'}
    ${false} | ${'hides'}
  `('$label when loading icon $loading is true', ({ loading }) => {
    factory(createCommitData(), loading);

    return vm.vm.$nextTick(() => {
      expect(vm.find(GlLoadingIcon).exists()).toBe(loading);
    });
  });

  it('renders commit widget', () => {
    factory();

    return vm.vm.$nextTick(() => {
      expect(vm.element).toMatchSnapshot();
    });
  });

  it('renders short commit ID', () => {
    factory();

    return vm.vm.$nextTick(() => {
      expect(vm.find('.label-monospace').text()).toEqual('12345678');
    });
  });

  it('hides pipeline components when pipeline does not exist', () => {
    factory(createCommitData({ pipeline: null }));

    return vm.vm.$nextTick(() => {
      expect(vm.find('.js-commit-pipeline').exists()).toBe(false);
    });
  });

  it('renders pipeline components', () => {
    factory();

    return vm.vm.$nextTick(() => {
      expect(vm.find('.js-commit-pipeline').exists()).toBe(true);
    });
  });

  it('hides author component when author does not exist', () => {
    factory(createCommitData({ author: null }));

    return vm.vm.$nextTick(() => {
      expect(vm.find('.js-user-link').exists()).toBe(false);
      expect(vm.find(UserAvatarLink).exists()).toBe(false);
    });
  });

  it('does not render description expander when description is null', () => {
    factory(createCommitData({ description: null }));

    return vm.vm.$nextTick(() => {
      expect(vm.find('.text-expander').exists()).toBe(false);
      expect(vm.find('.commit-row-description').exists()).toBe(false);
    });
  });

  it('expands commit description when clicking expander', () => {
    factory(createCommitData({ description: 'Test description' }));

    return vm.vm
      .$nextTick()
      .then(() => {
        vm.find('.text-expander').vm.$emit('click');
        return vm.vm.$nextTick();
      })
      .then(() => {
        expect(vm.find('.commit-row-description').isVisible()).toBe(true);
        expect(vm.find('.text-expander').classes('open')).toBe(true);
      });
  });

  it('renders the signature HTML as returned by the backend', () => {
    factory(createCommitData({ signatureHtml: '<button>Verified</button>' }));

    return vm.vm.$nextTick().then(() => {
      expect(vm.element).toMatchSnapshot();
    });
  });

  it('sets correct CSS class if the commit message is empty', () => {
    factory(createCommitData({ message: '' }));

    return vm.vm.$nextTick().then(() => {
      expect(vm.find('.item-title').classes()).toContain(emptyMessageClass);
    });
  });
});
