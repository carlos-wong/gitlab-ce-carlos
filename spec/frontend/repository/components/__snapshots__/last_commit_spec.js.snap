// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository last commit component renders commit widget 1`] = `
<div
  class="info-well d-none d-sm-flex project-last-commit commit p-3"
>
  <user-avatar-link-stub
    class="avatar-cell"
    imgalt=""
    imgcssclasses=""
    imgsize="40"
    imgsrc="https://test.com"
    linkhref="https://test.com/test"
    tooltipplacement="top"
    tooltiptext=""
    username=""
  />
   
  <div
    class="commit-detail flex-list"
  >
    <div
      class="commit-content qa-commit-content"
    >
      <gl-link-stub
        class="commit-row-message item-title"
        href="https://test.com/commit/123"
      >
        
          Commit title
        
      </gl-link-stub>
       
      <!---->
       
      <div
        class="committer"
      >
        <gl-link-stub
          class="commit-author-link js-user-link"
          href="https://test.com/test"
        >
          
            Test
          
        </gl-link-stub>
        
          authored
          
        <timeago-tooltip-stub
          cssclass=""
          time="2019-01-01"
          tooltipplacement="bottom"
        />
      </div>
       
      <!---->
    </div>
     
    <div
      class="commit-actions flex-row"
    >
      <!---->
       
      <div
        class="ci-status-link"
      >
        <gl-link-stub
          class="js-commit-pipeline"
          href="https://test.com/pipeline"
          title="Commit: failed"
        >
          <ci-icon-stub
            aria-label="Commit: failed"
            cssclasses=""
            size="24"
            status="[object Object]"
          />
        </gl-link-stub>
      </div>
       
      <div
        class="commit-sha-group d-flex"
      >
        <div
          class="label label-monospace monospace"
        >
          
            12345678
          
        </div>
         
        <clipboard-button-stub
          cssclass="btn-default"
          text="123456789"
          title="Copy commit SHA"
          tooltipplacement="bottom"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Repository last commit component renders the signature HTML as returned by the backend 1`] = `
<div
  class="info-well d-none d-sm-flex project-last-commit commit p-3"
>
  <user-avatar-link-stub
    class="avatar-cell"
    imgalt=""
    imgcssclasses=""
    imgsize="40"
    imgsrc="https://test.com"
    linkhref="https://test.com/test"
    tooltipplacement="top"
    tooltiptext=""
    username=""
  />
   
  <div
    class="commit-detail flex-list"
  >
    <div
      class="commit-content qa-commit-content"
    >
      <gl-link-stub
        class="commit-row-message item-title"
        href="https://test.com/commit/123"
      >
        
          Commit title
        
      </gl-link-stub>
       
      <!---->
       
      <div
        class="committer"
      >
        <gl-link-stub
          class="commit-author-link js-user-link"
          href="https://test.com/test"
        >
          
            Test
          
        </gl-link-stub>
        
          authored
          
        <timeago-tooltip-stub
          cssclass=""
          time="2019-01-01"
          tooltipplacement="bottom"
        />
      </div>
       
      <!---->
    </div>
     
    <div
      class="commit-actions flex-row"
    >
      <div>
        <button>
          Verified
        </button>
      </div>
       
      <div
        class="ci-status-link"
      >
        <gl-link-stub
          class="js-commit-pipeline"
          href="https://test.com/pipeline"
          title="Commit: failed"
        >
          <ci-icon-stub
            aria-label="Commit: failed"
            cssclasses=""
            size="24"
            status="[object Object]"
          />
        </gl-link-stub>
      </div>
       
      <div
        class="commit-sha-group d-flex"
      >
        <div
          class="label label-monospace monospace"
        >
          
            12345678
          
        </div>
         
        <clipboard-button-stub
          cssclass="btn-default"
          text="123456789"
          title="Copy commit SHA"
          tooltipplacement="bottom"
        />
      </div>
    </div>
  </div>
</div>
`;
